include/tudat/astro/system_models/selfShadowing.h

```diff
@@ -0,0 +1,491 @@
+                std::floor((projections_[index][j].getMinL( )-minL)/sizeL)) - 1, 0, static_cast< int >( numberL ) - 1);
+            int indexMaxL = clamp(static_cast< int >( 
+                std::ceil((projections_[index][j].getMaxL( )-minL)/sizeL)) + 1, indexMinL + 1, static_cast< int >( numberL ));
+            int indexMinM = clamp(static_cast< int >(
+                std::floor((projections_[index][j].getMinM( )-minM)/sizeM)) - 1, 0, static_cast< int >( numberM ) - 1);
+            int indexMaxM = clamp(static_cast< int >(
+                std::ceil((projections_[index][j].getMaxM( )-minM)/sizeM)) + 1, indexMinM + 1, static_cast< int >( numberM ));
+            // apply PIP to shadowing and update pixelation matrix (assign value of 1)  
+            pixelationMatrix = arePointsInTriangle(projections_[index][j], 
+                gridCoordinatesL, gridCoordinatesM, pixelationMatrix, indexMinL, indexMaxL, indexMinM, indexMaxM);
+        }
+        double fractionOnes = 0;
+        for (int p = 0; p< static_cast< int >( gridCoordinatesM.size() ); p++) 
+        {
+            for (int q = 0; q< static_cast< int >( gridCoordinatesL.size() ); q++) 
+            {
+                if (pixelationMatrix[p][q] == 1) 
+                {
+                    fractionOnes++;
+                }
+            }
+        }
+        double fractionZeros = 0;
+        for (int p = 0; p< static_cast< int >( gridCoordinatesM.size() ); p++) 
+        {
+            for (int q = 0; q< static_cast< int >( gridCoordinatesL.size() ); q++) 
+            {
+                if (pixelationMatrix[p][q] == 0) 
+                {
+                    fractionZeros++;
+                }
+            }
+        }
+        
+        fractions[i] = clamp( 1 - fractionOnes / (fractionOnes + fractionZeros ), 0.0, 1.0 );
+    }
+    return fractions;
+}
+
+class SelfShadowing
+{
+private:
+    std::vector< double > illuminatedPanelFractions_;
+public:
+    SelfShadowing( ) : illuminatedPanelFractions_( ) { };
+
+    SelfShadowing( const std::vector< std::shared_ptr< system_models::VehicleExteriorPanel > >& allPanels,
+                   const Eigen::Vector3d& incomingDirection, 
+                   const int maximumNumberOfPixels,
+                   const bool multiThreading = false )
+    { 
+        int numberOfPanels = allPanels.size( );
+        // initialize sigma matrix for discrimination logic (-2 dummy value, -1 no-sh, 0 partial sh, 1 full sh)
+        std::vector<std::vector< int >> sigmaMatrix(numberOfPanels, std::vector< int >(numberOfPanels, -2));
+        // first discrimination
+        std::vector< int > firstDiscriminationIndexes;
+        for (int i = 0; i < numberOfPanels; i++) {
+            if ( !firstDiscrimination( allPanels[i], incomingDirection ) ){
+                firstDiscriminationIndexes.push_back( i );
+                for (int j = 0; j < numberOfPanels; j++) {
+                    sigmaMatrix[j][i] = -1; // cannot shadow other plates

```

Comment:
for accessing std::vector elements, used .at( i ) rather than [i]. Using [i] can lead to memory corruption/segfaults if used out of bounds.